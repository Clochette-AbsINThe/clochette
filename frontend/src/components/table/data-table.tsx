import { useState } from 'react';

import { ColumnDef, SortingState, VisibilityState, flexRender, getCoreRowModel, getFacetedRowModel, getFacetedUniqueValues, getFilteredRowModel, getPaginationRowModel, getSortedRowModel, useReactTable } from '@tanstack/react-table';

import { DataTablePagination } from '@/components/table/data-table-pagination';
import { Table, TableBody, TableCell, TableHead, TableHeader, TableRow } from '@/components/ui/table';

interface DataTableProps<TData extends { id: number }, TValue> {
  columns: ColumnDef<TData, TValue>[];
  data: TData[];
  pagination?: boolean;
}

export function DataTable<TData extends { id: number }, TValue>({ columns, data, pagination = true }: Readonly<DataTableProps<TData, TValue>>) {
  const [rowSelection, setRowSelection] = useState({});
  const columnVisibility: VisibilityState = { id: false };
  const [sorting, setSorting] = useState<SortingState>([{ id: 'id', desc: true }]);

  const table = useReactTable({
    sortDescFirst: true,
    data,
    columns,
    state: {
      sorting,
      columnVisibility,
      rowSelection
    },
    enableRowSelection: true,
    onRowSelectionChange: setRowSelection,
    onSortingChange: setSorting,
    getCoreRowModel: getCoreRowModel(),
    getFilteredRowModel: getFilteredRowModel(),
    getPaginationRowModel: getPaginationRowModel(),
    getSortedRowModel: getSortedRowModel(),
    getFacetedRowModel: getFacetedRowModel(),
    getFacetedUniqueValues: getFacetedUniqueValues()
  });

  if (!pagination) {
    table.getState().pagination.pageSize = data.length;
  }

  return (
    <div className='space-y-4 lg:mx-8 mx-2'>
      <div className='rounded-md border'>
        <Table>
          <TableHeader>
            {table.getHeaderGroups().map((headerGroup) => (
              <TableRow key={headerGroup.id}>
                {headerGroup.headers.map((header) => {
                  return <TableHead key={header.id}>{header.isPlaceholder ? null : flexRender(header.column.columnDef.header, header.getContext())}</TableHead>;
                })}
              </TableRow>
            ))}
          </TableHeader>
          <TableBody>
            {table.getRowModel().rows?.length ? (
              table.getRowModel().rows.map((row) => (
                <TableRow key={row.id}>
                  {row.getVisibleCells().map((cell) => (
                    <TableCell key={cell.id}>{flexRender(cell.column.columnDef.cell, cell.getContext())}</TableCell>
                  ))}
                </TableRow>
              ))
            ) : (
              <TableRow>
                <TableCell
                  colSpan={columns.length}
                  className='h-24 text-center'
                >
                  No results.
                </TableCell>
              </TableRow>
            )}
          </TableBody>
        </Table>
      </div>
      {pagination && <DataTablePagination table={table} />}
    </div>
  );
}
