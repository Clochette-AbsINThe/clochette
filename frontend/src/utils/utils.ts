import { AxiosResponse } from 'axios';
import { environmentVariable } from '@utils/settings';

export const addIdToUrl = (id: number): void => {
    const url = new URL(window.location.href);
    url.searchParams.set('id', id.toString());
    window.history.pushState({}, '', url.href);
};

export const removeIdFromUrl = (): void => {
    const url = new URL(window.location.href);
    url.searchParams.delete('id');
    window.history.pushState({}, '', url.href);
};

export const getIdFromUrl = (): number | null => {
    const url = new URL(window.location.href);
    const id = url.searchParams.get('id');
    if (id === null) return null;
    if (/^(-|)[0-9]+$/.test(id)) return parseInt(id);
    return null;
};

export function getErrorMessage(data: AxiosResponse<unknown, any>): string {
    if (data.status === 422) {
        const { detail } = data.data as {
            detail: [
                {
                    loc: [string, number];
                    msg: string;
                    type: string;
                }
            ];
        };
        return (
            detail
                .map((e) => e.msg)
                .join(' ')
                .charAt(0)
                .toUpperCase() +
            detail
                .map((e) => e.msg)
                .join(' ')
                .slice(1)
        );
    } else if (data.status === 500) {
        return 'Erreur serveur';
    } else {
        const { detail } = data.data as { detail: string };
        return detail;
    }
}

export interface Token {
    sub: string;
    iat: number;
    exp: number;
    token_type: string;
}

/**
 * This function could throw an error if the token is invalid in its format, for example if the token is not a JWT token
 * @param token JWT token
 * @returns Inforamtions inside the token
 */
export function parseJwt(token: string): Token {
    const base64Url = token.split('.')[1];
    const base64 = base64Url?.replace(/-/g, '+').replace(/_/g, '/');
    const buffer = Buffer.from(base64 as string, 'base64');
    const jsonPayload = decodeURIComponent(
        buffer
            .toString('utf-8')
            .split('')
            .map((c) => '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2))
            .join('')
    );
    return JSON.parse(jsonPayload);
}

export function getRedirectUrlEncoded(url: string): string {
    return encodeURIComponent(environmentVariable.BASE_URL + url);
}
