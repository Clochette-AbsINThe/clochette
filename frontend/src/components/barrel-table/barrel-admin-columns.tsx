import { ColumnDef, Row } from '@tanstack/react-table';

import { DataTableColumnHeader } from '@/components/table/data-table-column-header';
import { useReadGlasses } from '@/openapi-codegen/clochetteComponents';
import { Barrel } from '@/openapi-codegen/clochetteSchemas';
import { DecreasingArrowIcon, IncreaseArrowIcon } from '@/styles/utils';
import { formatPrice } from '@/utils/utils';

export const barrelColumns: ColumnDef<Barrel>[] = [
  {
    id: 'id',
    accessorKey: 'id'
  },
  {
    id: 'name',
    accessorKey: 'name',
    header: ({ column }) => (
      <DataTableColumnHeader
        column={column}
        title='Nom'
      />
    ),
    cell: ({ row }) => {
      const name = row.getValue<string>('name');

      if (row.original.emptyOrSolded) {
        return <span className='line-through'>{name}</span>;
      } else {
        return <span>{name}</span>;
      }
    }
  },
  {
    id: 'isMounted',
    accessorKey: 'isMounted',
    header: ({ column }) => (
      <DataTableColumnHeader
        column={column}
        title='Etat'
      />
    ),
    cell: ({ row }) => {
      const isMounted = row.getValue<boolean>('isMounted');

      return isMounted ? 'MontÃ© sur une tireuse' : 'Vide';
    },
    enableSorting: false
  },
  {
    id: 'buyPrice',
    accessorKey: 'buyPrice',
    header: ({ column }) => (
      <DataTableColumnHeader
        column={column}
        title="Prix d'achat"
      />
    ),
    cell: ({ row }) => {
      const amount = parseFloat(row.getValue('buyPrice'));
      const formatted = formatPrice(amount);

      return formatted;
    }
  },
  {
    id: 'totalGlassSold',
    header: ({ column }) => (
      <DataTableColumnHeader
        column={column}
        title='Nombre de verres vendus'
      />
    ),
    cell: ({ row }) => <TotalGlassSoldCell row={row} />
  },
  {
    id: 'totalSellPrice',
    header: ({ column }) => (
      <DataTableColumnHeader
        column={column}
        title='Prix total de vente'
      />
    ),
    cell: ({ row }) => <TotalSellPriceCell row={row} />
  },
  {
    id: 'profit',
    header: ({ column }) => (
      <DataTableColumnHeader
        column={column}
        title='Bilan'
      />
    ),
    cell: ({ row }) => <ProfitCell row={row} />
  }
];

export function TotalGlassSoldCell({ row }: Readonly<{ row: Row<Barrel> }>) {
  const { data } = useReadGlasses({
    queryParams: {
      barrel_id: row.getValue<number>('id')
    }
  });

  return <span>{data?.length ?? 0}</span>;
}

export function TotalSellPriceCell({ row }: Readonly<{ row: Row<Barrel> }>) {
  const { data } = useReadGlasses({
    queryParams: {
      barrel_id: row.getValue<number>('id')
    }
  });

  const totalSellPrice = data?.reduce((acc, glass) => {
    return acc + glass.sellPrice;
  }, 0);

  return <span>{formatPrice(totalSellPrice ?? 0)}</span>;
}

export function ProfitCell({ row }: Readonly<{ row: Row<Barrel> }>) {
  const { data } = useReadGlasses({
    queryParams: {
      barrel_id: row.getValue<number>('id')
    }
  });

  const totalSellPrice =
    data?.reduce((acc, glass) => {
      return acc + glass.sellPrice;
    }, 0) ?? 0;

  const profit = totalSellPrice - row.getValue<number>('buyPrice');

  return (
    <div className='flex justify-between'>
      <span className='mr-2'>{formatPrice(profit)}</span>
      {profit > 0 ? <IncreaseArrowIcon /> : <DecreasingArrowIcon />}
    </div>
  );
}
