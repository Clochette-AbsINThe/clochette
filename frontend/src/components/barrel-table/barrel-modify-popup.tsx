import { ResolverResult, useForm } from 'react-hook-form';
import { toast } from 'react-hot-toast';

import { useQueryClient } from '@tanstack/react-query';

import { DataTableRowActionsProps } from './barrel-table-row-actions';

import { Button } from '@/components/button';
import { DialogHeader, DialogTitle, DialogFooter } from '@/components/ui/dialog';
import { Form, FormControl, FormDescription, FormField, FormItem, FormLabel, FormMessage } from '@/components/ui/form';
import { Input } from '@/components/ui/input';
import { useReadBarrels, useUpdateBarrel } from '@/openapi-codegen/clochetteComponents';
import { generateApiErrorMessage } from '@/openapi-codegen/clochetteFetcher';
import { BarrelDistinct, BarrelUpdateModify } from '@/openapi-codegen/clochetteSchemas';
import { formatPrice } from '@/utils/utils';

export function barrelUpdateResolver(data: BarrelUpdateModify): ResolverResult<BarrelUpdateModify> {
  if (!data.sellPrice) {
    return {
      values: data,
      errors: {
        sellPrice: {
          type: 'required',
          message: 'Le prix de vente est requis.'
        }
      }
    };
  }
  if (data.sellPrice <= 0) {
    return {
      values: data,
      errors: {
        sellPrice: {
          type: 'min',
          message: 'Le prix de vente doit être supérieur à 0.'
        }
      }
    };
  }
  return {
    values: data,
    errors: {}
  };
}
interface BarrelModifyPopupProps {
  rowBarrel: BarrelDistinct;
  isOpen: boolean;
  setIsOpen: (isOpen: boolean) => void;
}

export function BarrelModifyPopup({ rowBarrel, isOpen, setIsOpen }: BarrelModifyPopupProps) {
  const queryClient = useQueryClient();

  const modifyBarrel = useUpdateBarrel({
    onError(error, variables, context) {
      const detail = generateApiErrorMessage(error);
      toast.error(`Erreur lors de la modification du fût. ${detail}`);
    }
  });

  const { data } = useReadBarrels(
    {
      queryParams: {
        drink_item_id: rowBarrel.drinkItemId
      }
    },
    {
      enabled: isOpen
    }
  );

  const form = useForm<BarrelUpdateModify>({
    defaultValues: {
      sellPrice: rowBarrel.sellPrice
    },
    resolver: barrelUpdateResolver
  });

  const onSubmit = (submitData: BarrelUpdateModify) => {
    data?.forEach((barrel) => {
      modifyBarrel.mutate(
        {
          body: submitData,
          pathParams: {
            barrelId: barrel.id
          }
        },
        {
          onSuccess(data, variables, context) {
            setIsOpen(false);
            toast.success(`Fût ${data.name} modifié avec succès !`);
            queryClient.invalidateQueries({ stale: true });
          }
        }
      );
    });
  };

  return (
    <>
      <DialogHeader>
        <DialogTitle>Modifier le fût :</DialogTitle>
      </DialogHeader>
      <div className='flex flex-col gap-4 min-h-[50vh] min-w-[50vw]'>
        {rowBarrel && (
          <>
            <div className='flex flex-col mt-2 gap-y-2'>
              <div className='flex items-center gap-x-4'>
                <h6 className='text-lg font-semi text-accent-foreground'>{rowBarrel.name}</h6>
              </div>
              <ul className='list-disc pl-6'>
                <li className='text-md text-muted-foreground'>Prix de vente au verre: {formatPrice(rowBarrel.sellPrice)}</li>
                <li className='text-md text-muted-foreground'>Prix d&apos;achat: {formatPrice(rowBarrel.buyPrice)}</li>
              </ul>
            </div>
            <div className='flex flex-col mt-2 gap-y-2'>
              <div className='flex items-center gap-x-4'>
                <h6 className='text-lg font-semi text-accent-foreground'>Modifier le fût :</h6>
              </div>
              <Form {...form}>
                <form
                  onSubmit={form.handleSubmit(onSubmit)}
                  className='space-y-8 flex flex-col flex-grow'
                >
                  <FormField
                    control={form.control}
                    name='sellPrice'
                    render={({ field }) => (
                      <FormItem>
                        <FormLabel>Prix de vente (€)</FormLabel>
                        <FormControl>
                          <Input
                            step='any'
                            type='number'
                            placeholder={formatPrice(rowBarrel.sellPrice)}
                            {...field}
                            value={field.value ?? 0}
                          />
                        </FormControl>
                        <FormDescription>Le prix de vente du verre associé à ce fût.</FormDescription>
                        <FormMessage />
                      </FormItem>
                    )}
                  />
                  <div className='grow'></div>
                  <DialogFooter>
                    <Button
                      confirm
                      loading={modifyBarrel.isLoading}
                      type='submit'
                    >
                      Modifier le fût
                    </Button>
                  </DialogFooter>
                </form>
              </Form>
            </div>
          </>
        )}
      </div>
    </>
  );
}
