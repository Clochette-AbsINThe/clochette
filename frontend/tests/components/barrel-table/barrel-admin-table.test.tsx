import { render } from '@testing-library/react';

import { barrelColumns } from '@/components/barrel-table/barrel-admin-columns';
import { DataTable } from '@/components/table/data-table';
import { Barrel } from '@/openapi-codegen/clochetteSchemas';
import { createWrapper, spyOnApiCall } from 'tests/utils';

const barrels: Barrel[] = [
  {
    name: 'Rouge',
    buyPrice: 65.0,
    sellPrice: 2.5,
    barrelSellPrice: null,
    isMounted: true,
    emptyOrSolded: true,
    drinkItemId: 1,
    id: 1
  },
  {
    name: 'Blanche',
    buyPrice: 80.0,
    sellPrice: 2.5,
    barrelSellPrice: null,
    isMounted: false,
    emptyOrSolded: false,
    drinkItemId: 1,
    id: 2
  }
];

describe('accountColumns', () => {
  it('should render all columns', () => {
    spyOnApiCall('useReadGlasses', [
      {
        sellPrice: 2.5
      }
    ]);

    const result = render(
      <DataTable
        columns={barrelColumns}
        data={barrels}
      />,
      { wrapper: createWrapper() }
    );

    expect(result.container).toMatchSnapshot();
  });
});
