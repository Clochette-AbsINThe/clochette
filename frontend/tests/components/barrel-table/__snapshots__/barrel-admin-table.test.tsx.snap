// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`accountColumns > should render all columns 1`] = `
<div>
  <div
    class="space-y-4 lg:mx-8 mx-2"
  >
    <div
      class="rounded-md border"
    >
      <div
        class="w-full overflow-auto"
      >
        <table
          class="w-full table-auto caption-bottom text-sm"
        >
          <thead
            class="[&_tr]:border-b"
          >
            <tr
              class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
            >
              <th
                class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
              >
                <div
                  class="flex items-center space-x-2"
                >
                  <button
                    aria-expanded="false"
                    aria-haspopup="menu"
                    class="inline-flex items-center justify-center font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground rounded-md px-3 text-xs -ml-3 h-8 data-[state=open]:bg-accent min-w-max"
                    data-state="closed"
                    id="radix-:r0:"
                    type="button"
                  >
                    <span>
                      Nom
                    </span>
                    <svg
                      class="ml-2 h-4 w-4"
                      fill="none"
                      height="15"
                      viewBox="0 0 15 15"
                      width="15"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M4.93179 5.43179C4.75605 5.60753 4.75605 5.89245 4.93179 6.06819C5.10753 6.24392 5.39245 6.24392 5.56819 6.06819L7.49999 4.13638L9.43179 6.06819C9.60753 6.24392 9.89245 6.24392 10.0682 6.06819C10.2439 5.89245 10.2439 5.60753 10.0682 5.43179L7.81819 3.18179C7.73379 3.0974 7.61933 3.04999 7.49999 3.04999C7.38064 3.04999 7.26618 3.0974 7.18179 3.18179L4.93179 5.43179ZM10.0682 9.56819C10.2439 9.39245 10.2439 9.10753 10.0682 8.93179C9.89245 8.75606 9.60753 8.75606 9.43179 8.93179L7.49999 10.8636L5.56819 8.93179C5.39245 8.75606 5.10753 8.75606 4.93179 8.93179C4.75605 9.10753 4.75605 9.39245 4.93179 9.56819L7.18179 11.8182C7.35753 11.9939 7.64245 11.9939 7.81819 11.8182L10.0682 9.56819Z"
                        fill="currentColor"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </button>
                </div>
              </th>
              <th
                class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
              >
                <div
                  class=""
                >
                  Etat
                </div>
              </th>
              <th
                class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
              >
                <div
                  class="flex items-center space-x-2"
                >
                  <button
                    aria-expanded="false"
                    aria-haspopup="menu"
                    class="inline-flex items-center justify-center font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground rounded-md px-3 text-xs -ml-3 h-8 data-[state=open]:bg-accent min-w-max"
                    data-state="closed"
                    id="radix-:r2:"
                    type="button"
                  >
                    <span>
                      Prix d'achat
                    </span>
                    <svg
                      class="ml-2 h-4 w-4"
                      fill="none"
                      height="15"
                      viewBox="0 0 15 15"
                      width="15"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M4.93179 5.43179C4.75605 5.60753 4.75605 5.89245 4.93179 6.06819C5.10753 6.24392 5.39245 6.24392 5.56819 6.06819L7.49999 4.13638L9.43179 6.06819C9.60753 6.24392 9.89245 6.24392 10.0682 6.06819C10.2439 5.89245 10.2439 5.60753 10.0682 5.43179L7.81819 3.18179C7.73379 3.0974 7.61933 3.04999 7.49999 3.04999C7.38064 3.04999 7.26618 3.0974 7.18179 3.18179L4.93179 5.43179ZM10.0682 9.56819C10.2439 9.39245 10.2439 9.10753 10.0682 8.93179C9.89245 8.75606 9.60753 8.75606 9.43179 8.93179L7.49999 10.8636L5.56819 8.93179C5.39245 8.75606 5.10753 8.75606 4.93179 8.93179C4.75605 9.10753 4.75605 9.39245 4.93179 9.56819L7.18179 11.8182C7.35753 11.9939 7.64245 11.9939 7.81819 11.8182L10.0682 9.56819Z"
                        fill="currentColor"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </button>
                </div>
              </th>
              <th
                class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
              >
                <div
                  class=""
                >
                  Nombre de verres vendus
                </div>
              </th>
              <th
                class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
              >
                <div
                  class=""
                >
                  Prix total de vente
                </div>
              </th>
              <th
                class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
              >
                <div
                  class=""
                >
                  Bilan
                </div>
              </th>
            </tr>
          </thead>
          <tbody
            class="[&_tr:last-child]:border-0"
          >
            <tr
              class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
            >
              <td
                class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
              >
                <span>
                  Blanche
                </span>
              </td>
              <td
                class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
              >
                Vide
              </td>
              <td
                class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
              >
                80,00 €
              </td>
              <td
                class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
              >
                <span>
                  1
                </span>
              </td>
              <td
                class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
              >
                <span>
                  2,50 €
                </span>
              </td>
              <td
                class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
              >
                <div
                  class="flex justify-between"
                >
                  <span
                    class="mr-2"
                  >
                    -77,50 €
                  </span>
                  <svg
                    class="text-red-600 h-4 w-4"
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <polyline
                      points="23 18 13.5 8.5 8.5 13.5 1 6"
                    />
                    <polyline
                      points="17 18 23 18 23 12"
                    />
                  </svg>
                </div>
              </td>
            </tr>
            <tr
              class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
            >
              <td
                class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
              >
                <span
                  class="line-through"
                >
                  Rouge
                </span>
              </td>
              <td
                class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
              >
                Monté sur une tireuse
              </td>
              <td
                class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
              >
                65,00 €
              </td>
              <td
                class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
              >
                <span>
                  1
                </span>
              </td>
              <td
                class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
              >
                <span>
                  2,50 €
                </span>
              </td>
              <td
                class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
              >
                <div
                  class="flex justify-between"
                >
                  <span
                    class="mr-2"
                  >
                    -62,50 €
                  </span>
                  <svg
                    class="text-red-600 h-4 w-4"
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <polyline
                      points="23 18 13.5 8.5 8.5 13.5 1 6"
                    />
                    <polyline
                      points="17 18 23 18 23 12"
                    />
                  </svg>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="flex items-center justify-end px-2 flex-wrap gap-2"
    >
      <div
        class="flex items-center gap-x-6 lg:gap-x-8 flex-wrap"
      >
        <div
          class="md:flex items-center gap-2 hidden"
        >
          <p
            class="text-sm font-medium"
          >
            Lignes par page
          </p>
          <button
            aria-autocomplete="none"
            aria-controls="radix-:r4:"
            aria-expanded="false"
            class="flex items-center justify-between rounded-md border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50 h-8 w-16"
            data-state="closed"
            dir="ltr"
            role="combobox"
            type="button"
          >
            <span
              style="pointer-events: none;"
            >
              10
            </span>
            <svg
              aria-hidden="true"
              class="h-4 w-4 opacity-50"
              fill="none"
              height="15"
              viewBox="0 0 15 15"
              width="15"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M4.93179 5.43179C4.75605 5.60753 4.75605 5.89245 4.93179 6.06819C5.10753 6.24392 5.39245 6.24392 5.56819 6.06819L7.49999 4.13638L9.43179 6.06819C9.60753 6.24392 9.89245 6.24392 10.0682 6.06819C10.2439 5.89245 10.2439 5.60753 10.0682 5.43179L7.81819 3.18179C7.73379 3.0974 7.61933 3.04999 7.49999 3.04999C7.38064 3.04999 7.26618 3.0974 7.18179 3.18179L4.93179 5.43179ZM10.0682 9.56819C10.2439 9.39245 10.2439 9.10753 10.0682 8.93179C9.89245 8.75606 9.60753 8.75606 9.43179 8.93179L7.49999 10.8636L5.56819 8.93179C5.39245 8.75606 5.10753 8.75606 4.93179 8.93179C4.75605 9.10753 4.75605 9.39245 4.93179 9.56819L7.18179 11.8182C7.35753 11.9939 7.64245 11.9939 7.81819 11.8182L10.0682 9.56819Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
        <div
          class="flex w-20 items-center justify-center text-sm font-medium"
        >
          Page 
          1
           sur 
          1
        </div>
        <div
          class="flex items-center gap-2"
        >
          <button
            class="items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-transparent shadow-sm hover:bg-accent hover:text-accent-foreground hidden h-8 w-8 p-0 lg:flex"
            disabled=""
          >
            <span
              class="sr-only"
            >
              Go to first page
            </span>
            <svg
              class="h-4 w-4"
              fill="none"
              height="15"
              viewBox="0 0 15 15"
              width="15"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M6.85355 3.85355C7.04882 3.65829 7.04882 3.34171 6.85355 3.14645C6.65829 2.95118 6.34171 2.95118 6.14645 3.14645L2.14645 7.14645C1.95118 7.34171 1.95118 7.65829 2.14645 7.85355L6.14645 11.8536C6.34171 12.0488 6.65829 12.0488 6.85355 11.8536C7.04882 11.6583 7.04882 11.3417 6.85355 11.1464L3.20711 7.5L6.85355 3.85355ZM12.8536 3.85355C13.0488 3.65829 13.0488 3.34171 12.8536 3.14645C12.6583 2.95118 12.3417 2.95118 12.1464 3.14645L8.14645 7.14645C7.95118 7.34171 7.95118 7.65829 8.14645 7.85355L12.1464 11.8536C12.3417 12.0488 12.6583 12.0488 12.8536 11.8536C13.0488 11.6583 13.0488 11.3417 12.8536 11.1464L9.20711 7.5L12.8536 3.85355Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <button
            class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-transparent shadow-sm hover:bg-accent hover:text-accent-foreground h-8 w-8 p-0"
            disabled=""
          >
            <span
              class="sr-only"
            >
              Go to previous page
            </span>
            <svg
              class="h-4 w-4"
              fill="none"
              height="15"
              viewBox="0 0 15 15"
              width="15"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M8.84182 3.13514C9.04327 3.32401 9.05348 3.64042 8.86462 3.84188L5.43521 7.49991L8.86462 11.1579C9.05348 11.3594 9.04327 11.6758 8.84182 11.8647C8.64036 12.0535 8.32394 12.0433 8.13508 11.8419L4.38508 7.84188C4.20477 7.64955 4.20477 7.35027 4.38508 7.15794L8.13508 3.15794C8.32394 2.95648 8.64036 2.94628 8.84182 3.13514Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <button
            class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-transparent shadow-sm hover:bg-accent hover:text-accent-foreground h-8 w-8 p-0"
            disabled=""
          >
            <span
              class="sr-only"
            >
              Go to next page
            </span>
            <svg
              class="h-4 w-4"
              fill="none"
              height="15"
              viewBox="0 0 15 15"
              width="15"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M6.1584 3.13508C6.35985 2.94621 6.67627 2.95642 6.86514 3.15788L10.6151 7.15788C10.7954 7.3502 10.7954 7.64949 10.6151 7.84182L6.86514 11.8418C6.67627 12.0433 6.35985 12.0535 6.1584 11.8646C5.95694 11.6757 5.94673 11.3593 6.1356 11.1579L9.565 7.49985L6.1356 3.84182C5.94673 3.64036 5.95694 3.32394 6.1584 3.13508Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <button
            class="items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-transparent shadow-sm hover:bg-accent hover:text-accent-foreground hidden h-8 w-8 p-0 lg:flex"
            disabled=""
          >
            <span
              class="sr-only"
            >
              Go to last page
            </span>
            <svg
              class="h-4 w-4"
              fill="none"
              height="15"
              viewBox="0 0 15 15"
              width="15"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M2.14645 11.1464C1.95118 11.3417 1.95118 11.6583 2.14645 11.8536C2.34171 12.0488 2.65829 12.0488 2.85355 11.8536L6.85355 7.85355C7.04882 7.65829 7.04882 7.34171 6.85355 7.14645L2.85355 3.14645C2.65829 2.95118 2.34171 2.95118 2.14645 3.14645C1.95118 3.34171 1.95118 3.65829 2.14645 3.85355L5.79289 7.5L2.14645 11.1464ZM8.14645 11.1464C7.95118 11.3417 7.95118 11.6583 8.14645 11.8536C8.34171 12.0488 8.65829 12.0488 8.85355 11.8536L12.8536 7.85355C13.0488 7.65829 13.0488 7.34171 12.8536 7.14645L8.85355 3.14645C8.65829 2.95118 8.34171 2.95118 8.14645 3.14645C7.95118 3.34171 7.95118 3.65829 8.14645 3.85355L11.7929 7.5L8.14645 11.1464Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
